### PR TITLE
Tests to fake odd screen metrics

### LINKED ISSUE #5: Recording does not work if screen height is odd
```
20:45:18.435 ERROR [VideoRec]      - [libx264 @ 00000000003DB9A0] height not divisible by 2 (1920x975)
20:45:18.492 ERROR [VideoRec]      - FfmpegException::make(finalStr): could not open codec: Error number -1 occurred (-1) (VideoExceptions.cpp:118)
20:45:18.588 ERROR [VideoRec]      - Video recorder encountered exception. Recording has been stopped.
java.lang.RuntimeException: could not open codec: Error number -1 occurred (-1)
	at io.humble.video.VideoJNI.Encoder_open(Native Method) ~[humble-video-noarch-0.2.1.jar:0.2.1]
	at io.humble.video.Encoder.open(Encoder.java:154) ~[humble-video-noarch-0.2.1.jar:0.2.1]
	at tdl.record.screen.video.VideoRecorder.open(VideoRecorder.java:118) ~[dev-screen-record-0.0.6.jar:na]
	at tdl.record_upload.video.VideoRecordingThread.run(VideoRecordingThread.java:35) ~[record-and-upload-0.0.16.jar:na]
```

### PR DESCRIPTION
**Pull request by neomatrix369**
- Tests input from screen input type
- Tests Input from stream bar codes input type

Fixes #5

### CODE REVIEW

**Comment by neomatrix369** [comment]
Can't see on the Codeship console why the build is failing, I get this screen:

image

Passes on my local machine:

```
± ./gradlew clean test                                                                                  
:clean
:compileJava
:processResources UP-TO-DATE
:classes
:compileTestJava
:processTestResources
:testClasses
:test
GraalVM 1.0.0-rc9 warning: You have loaded library /tmp/humble5884629870715503086.tmp which might have disabled stack guard. The VM will try to fix the stack guard now.
It's highly recommended that you fix the library with 'execstack -c <libfile>', or link it with '-z noexecstack'.

BUILD SUCCESSFUL

Total time: 34.19 secs

```

**Comment by julianghionoiu** [comment]
I have given you access to the Codeship console.
By the looks of it that is an old test. 
```
acceptance.CanHandleFailuresTest > recorded_file_should_be_recoverable_after_sigkill FAILED
java.lang.RuntimeException: Error opening url: build/recording_interrupted_by_sigkill.mp4; : Invalid data found when processing input (-1094995529)
at io.humble.video.VideoJNI.Demuxer_open(Native Method)
at io.humble.video.Demuxer.open(Demuxer.java:201)
at tdl.record.screen.video.VideoPlayer.open(VideoPlayer.java:50)
at acceptance.CanHandleFailuresTest.recorded_file_should_be_recoverable_after_sigkill(CanHandleFailuresTest.java:67)
```
The implementation should have covered SIGKILL on all cases but it looks like there might be some missing ones. This is a separate issue.

File: src/test/java/input_from_screen/CanVideoRecorderPlayWithDiffScreenMetrics.java [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+    private static final int ODD_WIDTH = 301;
+    private static final int EVEN_HEIGHT = 150;
+
+    @Test public void
+    given_screen_width_is_odd_adjust_width_and_video_should_still_play_fine() {
+        try {
+            Rectangle screenBounds = new Rectangle(ODD_WIDTH, EVEN_HEIGHT);
+            ImageInput imageInput = new ScaleToOptimalSizeImage(
+                    ImageQualityHint.HIGH,
+                    new InputFromScreen(screenBounds)
```
Comment (julianghionoiu) [inline src/test/java/input_from_screen/CanVideoRecorderPlayWithDiffScreenMetrics.java]: Use InputFromStaticImage instead, as InputFromScreen will not work on the CI (no screen).

File: src/test/java/input_from_screen/CanVideoRecorderPlayWithDiffScreenMetrics.java [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+        } catch (Exception ex) {
+            fail("Video should play perfectly fine, instead threw an exception: " + ex.getMessage());
+        }
+    }
+
+    @Test public void
+    given_screen_height_and_width_are_even_and_video_should_play_fine() {
+        try {
+            Rectangle screenBounds = new Rectangle(EVEN_WIDTH, EVEN_HEIGHT);
+            ImageInput imageInput = new ScaleToOptimalSizeImage(ImageQualityHint.MEDIUM, new InputFromScreen(screenBounds));
```
Comment (julianghionoiu) [inline src/test/java/input_from_screen/CanVideoRecorderPlayWithDiffScreenMetrics.java]: Use InputFromStaticImage instead, as InputFromScreen will not work on the CI (no screen).

File: src/test/java/input_from_screen/CanVideoRecorderPlayWithDiffScreenMetrics.java [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+        } catch (Exception ex) {
+            fail("Video should play perfectly fine, instead threw an exception: " + ex.getMessage());
+        }
+    }
+
+    @Test public void
+    given_screen_height_is_odd_adjust_height_and_video_should_still_play_fine() {
+        try {
+            Rectangle screenBounds = new Rectangle(EVEN_WIDTH, ODD_HEIGHT);
+            ImageInput imageInput = new ScaleToOptimalSizeImage(ImageQualityHint.LOW, new InputFromScreen(screenBounds));
```
Comment (julianghionoiu) [inline src/test/java/input_from_screen/CanVideoRecorderPlayWithDiffScreenMetrics.java]: Use InputFromStaticImage instead, as InputFromScreen will not work on the CI (no screen).

File: src/test/java/tdl/record/screen/image/input/InputFromStreamOfBarcodes.java [hunk 4]
```diff
@@ -22,9 +22,11 @@
     private long systemStartTime;
 
     public InputFromStreamOfBarcodes(BarcodeFormat barcodeFormat, int width, int height, TimeSource timeSource) {
-        this.width = width;
-        this.height = height;
-        this.barcodeImageRaster = new BarcodeImageRaster(width, height, this.height / 3);
+        this.width = convertToEvenNumber(width);
```
Comment (julianghionoiu) [inline src/test/java/tdl/record/screen/image/input/InputFromStreamOfBarcodes.java]: You do not need to change existing classes.

File: src/main/java/tdl/record/screen/image/input/InputFromScreen.java [hunk 5]
```diff
@@ -15,6 +15,15 @@ public InputFromScreen() {
         this.robot = null;
     }
 
+    public InputFromScreen(Rectangle screenBounds) throws InputImageGenerationException {
```
Comment (julianghionoiu) [inline src/main/java/tdl/record/screen/image/input/InputFromScreen.java:18]: You do not need to change existing classes.

File: src/test/java/acceptance/input_from_stream_of_barcodes/CanVideoRecorderPlayWithDiffScreenMetrics.java [hunk 6]
```diff
@@ -0,0 +1,76 @@
...
+import tdl.record.screen.time.FakeTimeSource;
+import tdl.record.screen.time.TimeSource;
+import tdl.record.screen.video.VideoRecorder;
+
+import java.time.Duration;
+import java.time.temporal.ChronoUnit;
+
+import static junit.framework.TestCase.fail;
+
+public class CanVideoRecorderPlayWithDiffScreenMetrics {
```
Comment (julianghionoiu) [inline src/test/java/acceptance/input_from_stream_of_barcodes/CanVideoRecorderPlayWithDiffScreenMetrics.java]: There is no need for this test. The unit level test is enough to cover this edge case.

File: src/test/java/input_from_screen/CanVideoRecorderPlayWithDiffScreenMetrics.java [hunk 7]
```diff
@@ -0,0 +1,79 @@
...
+import tdl.record.screen.utils.ImageQualityHint;
+import tdl.record.screen.video.VideoRecorder;
+
+import java.awt.*;
+import java.time.Duration;
+import java.time.temporal.ChronoUnit;
+
+import static junit.framework.TestCase.fail;
+
+public class CanVideoRecorderPlayWithDiffScreenMetrics {
```
Comment (julianghionoiu) [inline src/test/java/input_from_screen/CanVideoRecorderPlayWithDiffScreenMetrics.java:17]: This is a good test and should be kept with some modifications, see below.

File: src/main/java/tdl/record/screen/image/input/ImageInput.java [hunk 8]
```diff
@@ -22,4 +22,15 @@ BufferedImage readImage()
     int getHeight();
 
     void close();
+
+    default int convertToEvenNumber(int value) {
```
Comment (julianghionoiu) [inline src/main/java/tdl/record/screen/image/input/ImageInput.java]: You do not need to change existing classes.

**Review by julianghionoiu — COMMENTED** [review]
The unit test is ok. 
The acceptance test is not needed.
The implementation has to be implemented in a different way.

The current PR breaks encapsulation by changing the InputFromScreen class and the ImageInput.
The correct solution is to implement the behaviour into a separate class that implements ImageIn[ut and wraps another ImageInput. Similar in concept to ScaleToOptimalSizeImage. This is the Decorator pattern.

**Comment by neomatrix369** [comment]
Thanks for the useful feedback, it certainly felt wrong to do it this way - due to a couple of reasons, now you mention the alternatives I will use them instead.

But I wanted to get it to work and get it out in the whole first and get feedback.

**Comment by neomatrix369** [comment]
Have been able to reproduce the issue and also apply the fix in the last few commits

**Comment by neomatrix369** [comment]
Renaming and tidying up changes now and pushing commits

**Comment by julianghionoiu** [comment]
I have used the approach from your PR to create a separate commit:
https://github.com/julianghionoiu/dev-screen-record/commit/47f9d655631c0171e5e1ffe0ccb50c69444d6943

We can close this PR.
